### PR TITLE
AB testing framework

### PR DESCRIPTION
**Pull request by jranks123**
Implements AB testing framework, with an initial AA test implemented to test the framework.

### CODE REVIEW

File: frontend/app/controllers/Giraffe.scala [hunk 1]
```diff
@@ -25,12 +26,19 @@ object Giraffe extends Controller {
   val chargeId = "charge_id"
   val maxAmount: Option[Int] = 500.some
 
+
+  def createCookie(variant: TestTrait#Variant): Cookie = {
```
Comment (tomverran) [inline frontend/app/controllers/Giraffe.scala]: Maybe this function should be in the Test object in ABTest.scala?

File: frontend/app/forms/MemberForm.scala [hunk 2]
```diff
@@ -101,6 +105,16 @@ object MemberForm {
 
   case class FeedbackForm(category: String, page: String, feedback: String, name: String, email: String)
 
+  val abTestFormatter: Formatter[JsValue] = new Formatter[JsValue] {
+    override def bind(key: String, data: Map[String, String]): Either[Seq[FormError],JsValue] = {
+      printf(key)
```
Comment (tomverran) [inline frontend/app/forms/MemberForm.scala]: printf

File: frontend/app/forms/MemberForm.scala [hunk 3]
```diff
@@ -101,6 +105,16 @@ object MemberForm {
...
+      val parse: JsValue = Json.parse(URLDecoder.decode(data(key),"UTF8"))
```
Comment (tomverran) [inline frontend/app/forms/MemberForm.scala]: Maybe we could use a constant for UTF-8

http://stackoverflow.com/questions/6698354/where-to-get-utf-8-string-literal-in-java

File: frontend/app/views/fragments/giraffe/abFormTrack.scala.html [hunk 4]
```diff
@@ -0,0 +1,9 @@
+@import views.support.ChosenVariants
+@import java.net.URLEncoder
+@(chosenVariants: ChosenVariants)
+
+
+    <input type="hidden" name="abTest" class="js-ab-test" value="@URLEncoder.encode(chosenVariants.asJson,"UTF8")">
```
Comment (tomverran) [inline frontend/app/views/fragments/giraffe/abFormTrack.scala.html]: You could put 

URLEncoder.encode(chosenVariants.asJson,"UTF8")

as a function on chosenVariants for simplicity

File: frontend/app/forms/MemberForm.scala (hunk 2 above)
Comment (AWare) [inline frontend/app/forms/MemberForm.scala]: !(http://i.giphy.com/RIECDaCdxqKha.gif)[] 
This one was me.

Comment (jranks123) [inline frontend/app/forms/MemberForm.scala]: What do you mean @tomverran ?

File: frontend/app/forms/MemberForm.scala (hunk 3 above)
Comment (jranks123) [inline frontend/app/forms/MemberForm.scala]: eg StandardCharsets.UTF_8.name() ?

File: frontend/app/forms/MemberForm.scala (hunk 2 above)
Comment (jranks123) [inline frontend/app/forms/MemberForm.scala]: Ah ok i get it, dw

File: frontend/app/views/support/ABTest.scala [hunk 5]
```diff
@@ -49,7 +49,7 @@ object MessageCopyTest extends TestTrait {
 case class ChosenVariants(v1: AmountHighlightTest.Variant, v2: MessageCopyTest.Variant) {
   def asList: Seq[TestTrait#Variant] = Seq(v1,v2) //this makes me very sad
   def asJson = Json.toJson(asList).toString()
-  def encodeUTF8 = URLEncoder.encode(asJson, StandardCharsets.UTF_8.name())
+  def encodeURL = URLEncoder.encode(asJson, StandardCharsets.UTF_8.name())
```
Comment (AWare) [inline frontend/app/views/support/ABTest.scala]: 💯

**Comment by tomverran** [comment]
👍

**Pull request merged**